idioms/priv-extend.md

```diff
@@ -1,46 +1,103 @@
-# Privacy for extensibility
+# `#[non_exhaustive]` & private fields for extensibility
 
 ## Description
 
-Use a private field to ensure that a struct is extensible without breaking
-stability guarantees.
+A small set of scenarios exist where a library author may want to add public
+fields to a public struct or new variants to an enum without breaking backwards
+compatibility. Rust offers two solutions:
+
+- Use `#[non_exhaustive]` on `struct`s, `enum`s, and `enum` variants.
+For extensive documentation on all the places `#[non_exhaustive]` can be used,
+see [the docs](https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute).
+
+- For `struct`s only, you may add a private field to struct
+
+## Warning
+
+Use this deliberately and with caution: Incrementing the major version when adding
+fields or variants is often a better option. `#[non_exhaustive]` may be appropriate
+in scenarios where you're modeling an external resource that may change out-of-sync
+with your library, but is not a general purpose tool.
+
+`#[non_exhaustive]` forces clients to handle the "Something else" case; there is
+rarely a sensible action to take in this scenario. This leads to awkward code and
+code paths that only executed in extremely rare circumstances.
 
 ## Example
 
 ```rust,ignore
 mod a {
     // Public struct.
+    #[non_exhaustive]
     pub struct S {
         pub foo: i32,
-        // Private field.
-        bar: i32,
+    }
+    
+    #[non_exhaustive]
+    pub enum AdmitMoreVariants {
+        VariantA,
+        VariantB,
+        #[non_exhaustive]
+        VariantC { a: String }
     }
 }
 
+// Main function in another crate
 fn main(s: a::S) {
-    // Because S::bar is private, it cannot be named here and we must use `..`
-    // in the pattern.
+    // Because S is `#[non_exhaustive]`, it cannot be named here and
+    // we must use `..` in the pattern.
     let a::S { foo: _, ..} = s;
+    
+    let some_enum = a::AdmitMoreVariants::VariantA;
+    match some_enum {
+        a::AdmitMoreVariants::VariantA => println!("it's an A");
+        a::AdmitMoreVariants::VariantB => println!("it's a b");
+
+        // .. required because this variant is non-exhaustive as well
+        a::AdmitMoreVariants::VariantC { a, .. } => println!("it's a c");
+
+        // The wildcard match is required because more variants may be
+        // added in the future
+        _ => println1("it's a new variant");
+    }
 }
+```
+
```

Comment:
The example finishes here. We should move the rest of the section to another section or create a new section.